### DISCUSSION TITLE
Conditional depending on if user is editing in Elementor?

### DISCUSSION THREAD

**Discussion by stuartgreen-attain** [elementor/elementor#19556] [Developers Community]
I've done some research into this but the proposed solutions don't appear to be working for me, does anyone know if this has changed?

`\Elementor\Plugin::$instance->preview->is_preview_mode();`

I have tried this and other variations such as `\Elementor\Plugin::$instance->editor->is_edit_mode()` but I get the following error:

```
Fatal error: Uncaught Error: Call to a member function is_preview_mode() on null
```

I am using Elementor Version 3.7.2 and Elementor Pro Version 3.7.3

Is there any other option? I need to show/hide something based on if the user is editing in Elementor or not

**Comment by m1ga** [comment]
working fine here:

![Screenshot_20220822_134037](https://user-images.githubusercontent.com/4334997/185913006-dcc56382-558d-4a53-aca1-3d87ce49bb09.png)

(echo is for debugging of course)

**Reply by stuartgreen-attain** [reply]
![error-2](https://user-images.githubusercontent.com/98826108/185914380-d80ac7b8-d589-4828-b5f4-5c5954a59bb1.PNG)
![error-1](https://user-images.githubusercontent.com/98826108/185914386-564c078e-a616-4a08-bc3e-4341db0649f0.PNG)

Hmm how strange... I just copied your line 27 but I still get the same error... I guess I should try disabling themes/plugins?

**Reply by m1ga** [reply]
since $instance is null: is elementor active? And I'm not sure if `use Elementor\Plugin;` is required to make it work

**Comment by stuartgreen-attain** [comment]
Yeah Elementor and Elementor Pro are both active and this is what happens when I print_r the \Elementor\Plugin::$instance; by itself

On the frontend everything is empty, but when I edit a page with Elementor it doesn't load and goes into safe mode (and still doesn't load anything)

![screencapture-blgc-testing-attaindev-co-uk-2022-08-22-13_13_55](https://user-images.githubusercontent.com/98826108/185919261-7857dec1-6884-4a0b-8bbb-09b32d045951.png)

**Reply by m1ga** [reply]
thats strange. Even in frontend (private browsing) the print_r items are filled
![Screenshot_20220822_142331](https://user-images.githubusercontent.com/4334997/185920447-023c0253-43f4-42ab-8bc3-18fe762d0fe5.png)

and `[editor] => Elementor\Core\Editor\Editor Object..` or `[frontend] => Elementor\Frontend Object...`

**Reply by stuartgreen-attain** [reply]
Hm, well I'm at a loss 😂

**Reply by WebsiteDons** [reply]
If the environment has no access to the `$instance` variable, there will not be availability member function `is_edit_mode()`

Try using the method name `instance()` as `\Elementor\Plugin::instance()->editor->is_edit_mode()`

**Reply by stuartgreen-attain** [reply]
Hey, thanks for the response, I have just tried this but the output seems to be the same :-( 
![new-output](https://user-images.githubusercontent.com/98826108/187387335-d19e0edf-a6ee-48dc-a64b-97327ee3eeb1.PNG)

**Reply by WebsiteDons** [reply]
What result did you get when dump `\Elementor\Plugin::instance()->editor->is_edit_mode()`? It should be 1 if in edit view. Are you running the check in a widget or just checking to confirm in another plugin that the user is currently editing? If you define the complete process, or post the full code, the problem can be better diagnosed.

**Reply by stuartgreen-attain** [reply]
I get the same result, the instance itself is empty and if I dump the is_edit_mode I get the same `Uncaught Error: Call to a member function is_edit_mode() on null`

The process is, this line is in my functions.php file, where currently I am doing nothing other than dumping `$elementor_edit_active = \Elementor\Plugin::instance()->editor->is_edit_mode();` so there is no full code to post...

**Reply by WebsiteDons** [reply]
The method is purposed to run within the elementor environment so it has to be used in a class extension or abstract. It is not defined in the functions.php. 

example
```php
use Elementor\Widget_Base;

class My_Class extends Widget_Base {
    protected function render() {
     $mode =  \Elementor\Plugin::$instance->editor->is_edit_mode();
     $myAction='my action is running now because the mode is true';
     if( $mode ) 
        echo $myAction;
    }
}
```

**Reply by stuartgreen-attain** [reply]
Hm, I tried to put this in my functions file but this gives me the following error:

`Fatal error: Uncaught Error: Class 'Elementor\Widget_Base' not found in /var/www/vhosts/attaindev.co.uk/blgc.testing.attaindev.co.uk/wp-content/themes/generatepress-child/functions.php:25 Stack trace: #0`

```
use Elementor\Widget_Base;

class My_Class extends Widget_Base {
    protected function render() {
     $mode =  \Elementor\Plugin::$instance->editor->is_edit_mode();
     $myAction='my action is running now because the mode is true';
     if( $mode ) 
        echo $myAction;
    }
}

$test = new My_Class();
$test->render();
```

**Reply by WebsiteDons** [reply]
That code example was just that and never meant for you to use, I should have given more clarity. The template's functions.php is not within Elementor's environment therefore **you cannot call the methods using that file**.

You need to develop a plugin which **extends** Elementor's classes for various activities (widgets, modules, editor etc), then define the actions you need. See the developer guide at https://developers.elementor.com/docs/getting-started/first-addon/

You can check the functions.php for the loaded classes, methods, variable by dumping the following PHP methods. **Note: depending on your server's memory limits, the output attempt may cause error**.

```php
// test for loaded methods of Plugin class. There will be noting
print_r( get_class_methods('Plugin') );

// test for declared classes will show from all environments including PHP but you just need to look for any with Elementor/ 
print_r( get_declared_classes() );

// test for all variables loaded 
print_r( get_defined_vars() );

// test for variables of specific class
print_r( get_class_vars(get_class(new Plugin)) );
```

**Reply by m1ga** [reply]
ohhhh good point! I was thinking that @stuartgreen-attain is working in a plugin! That would explain why my code wasn't working for him (like the `use Elementor\Widget_Base;` part was already mentioned before) :rofl:
